### ISSUE TITLE
Support multiple Tags

### ISSUE THREAD

**Issue by jmgilman** [nlewo/nix2container#59]
Per the title, is it currently possible to provide multiple tags for an image? For example, tag it with the output hash and the git commit SHA simultaneously? Currently, it seems like you need two different derivations (identical at the binary level but configured with different tags at the manifest level). 

My current theory is to call `copyToDockerDaemon` and manually tag/publish the image myself, but it would be nice to be able to use `copyToRegistry` with multiple tags defined. Thoughts?

**Comment by nlewo** [comment]
@jmgilman Maybe it would be possible to build a script which calls `copyTo` with different destinations?

BTW, I don't think using the git commit ID is optimal since you can end up having the same image with a lot of labels. I suppose you want to know the commit ID to be able to rebuild an image which is deployed.
If you are using Hydra, it might be possible to search for the image output hash (the image tag) or use the [reproduce command of the hydra-cli client](https://github.com/nlewo/hydra-cli#reproduce) which i developed few years ago for this purpose.

**Comment by nrdxp** [comment]
I believe the idea, if Josh is building off of some of the stuff I showed him, would be to tag the image based on the Nix derivation input hash of its contents, since that's a pretty good tracker of if something _inside_ the image has actually changed, and then to incrementally add tags of the commit hashes that match this input hash so you have the metadata available to know which shas are identical. Is this correct?

Also, we have a lot of IFD in many of our projects due to using haskell.nix pretty extensively at work, so Hydra really isn't feasible for our CI.

**Comment by jmgilman** [comment]
@nrdxp  If I understand this correctly, I don't see a case where you could have the same git commit hash associated with more than one Nix output hash, _unless_ the `flake.nix` isn't being revision controlled. Correct? The commit hash should be unique across all changes to the underlying source code - so there should normally be a 1:1 relationship between the Nix output hash and the associated commit hash.

@nlewo You are correct in assuming this is to make finding images more accessible. Hydra isn't an option here.

**Comment by nrdxp** [comment]
> @nrdxp If I understand this correctly, I don't see a case where you could have the same git commit hash associated with more than one Nix output hash

It's totally possible, I'd share the code but its still private for a few more months, I believe I already sent them to you in Slack though. It depends on how you structure the image, and you have to be intentional about that, but with a decent source filter the input hash won't change until something in the source changes.

In Mamba, for example, we have a filter on the particular Scala modules in use by the project, so if nothing in those files changes, then input hash won't change. nix2container does CA tagging by default but that doesn't work in our case since the `self.rev` git revision is actually an input to the build.

So the hash I actually take is really a derivation of the package that is identical _sans_ `self.rev`. That way the revision is allowed to change freely while still giving me a good guage of actual changes to the contents otherwise.

So if you put it all together, you may have a few commits done to unrelated files in the source tree that would all produce the same output hash on your image, since nothing in _it_ changed. If we could associate every revision that matches a particular output hash, that'd be good meta-data to track.

**Comment by jmgilman** [comment]
Yes, I'm following now, took me a minute to wrap my head around it :)

**Comment by jmgilman** [comment]
@nlewo 

What if we turned `tag` -> `tags` and made it a list. Then, here:

```
copyToRegistry = image: pkgs.writeShellScriptBin "copy-to-registry" ''
    echo "Copy to Docker registry image ${image.imageName}:${image.imageTag}"
    ${skopeo-nix2container}/bin/skopeo --insecure-policy copy nix:${image} docker://${image.imageName}:${image.imageTag} $@
  '';
```

It would just loop through `tags` and call `skopeo copy` for each one?

**Comment by jmgilman** [comment]
Or, to maintain backwards compatibility, we add an `extraTags` input that is checked for in `copyToRegistry` and `skopeo` is invoked additional times for each extra tag.

**Comment by nlewo** [comment]
To be sure to understand, let me try to rephrase. 

You want to push your images with two tags: the image output hash (the tag by default) and a tag containing the commit ID of the flake. Is it correct?

**Comment by jmgilman** [comment]
In my case I'm not using the output hash at all, but yes that's the general premise. Images can have multiple tags applied, I'm simply proposing adding support for that fact. The alternative is copying to the local registry and manually publishing, which is a pretty bad UX if you ask me. The PR I submitted meets the above expectation.

**Comment by jmgilman** [comment]
Looks like there was some previous discussion on this here: https://github.com/containers/skopeo/issues/513

The latest version of skopeo seems to still suggest that this flag is limited to Docker only, so I'm not sure how feasible it is to use at this point.

**Comment by nlewo** [comment]
As previously discussed, i would prefer to avoid adding tags, names, and registries attributes to the buildImage function, since it's purpose it to build an image and not to push it.
I also understand you want to attach such kind information to the image. So, the [`meta` attribute](#65) looks like a good tradeoff :+1: 

I'm closing this issue since it has been addressed by #65
